azext_edge/edge/providers/orchestration/upgrade2.py

```diff
@@ -0,0 +1,309 @@
+
+from json import dumps
+from typing import Dict, List, Optional, Tuple
+
+from azure.cli.core.azclierror import ValidationError
+from knack.log import get_logger
+from packaging import version
+from rich.console import Console
+from rich.json import JSON
+from rich.progress import (
+    BarColumn,
+    Progress,
+    SpinnerColumn,
+    TextColumn,
+    TimeElapsedColumn,
+)
+from rich.table import Table, box
+
+from ...util import parse_kvp_nargs
+from ...util.common import should_continue_prompt
+from .common import (
+    EXTENSION_MONIKER_TO_ALIAS_MAP,
+    EXTENSION_TYPE_OPS,
+    EXTENSION_TYPE_TO_MONIKER_MAP,
+)
+from .resources import Instances
+from .targets import InitTargets
+
+logger = get_logger(__name__)
+
+DEFAULT_CONSOLE = Console()
+
+
+def upgrade_ops_instance(
+    cmd,
+    resource_group_name: str,
+    instance_name: str,
+    no_progress: Optional[bool] = None,
+    confirm_yes: Optional[bool] = None,
+    **kwargs,
+):
+    upgrade_manager = UpgradeManager(
+        cmd=cmd,
+        instance_name=instance_name,
+        resource_group_name=resource_group_name,
+        no_progress=no_progress,
+    )
+
+    upgrade_state = upgrade_manager.analyze_cluster(**kwargs)
+
+    if not no_progress:
+        render_upgrade_table(upgrade_state)
+
+    should_bail = not should_continue_prompt(confirm_yes=confirm_yes, context="Upgrade")
+    if should_bail:
+        return
+
+    if not upgrade_state.has_upgrades():
+        logger.warning("Nothing to upgrade :)")
+        return
+
+    return upgrade_manager.apply_upgrades(upgrade_state)
+
+
+class UpgradeManager:
+    def __init__(
+        self,
+        cmd,
+        resource_group_name: str,
+        instance_name: str,
+        no_progress: Optional[bool] = None,
+    ):
+        self.cmd = cmd
+        self.instance_name = instance_name
+        self.resource_group_name = resource_group_name
+        self.no_progress = no_progress
+        self.instances = Instances(self.cmd)
+        self.resource_map = self.instances.get_resource_map(
+            self.instances.show(name=self.instance_name, resource_group_name=self.resource_group_name)
+        )
+        self.targets = InitTargets(
+            cluster_name=self.resource_map.connected_cluster.cluster_name, resource_group_name=resource_group_name
+        )
+        self.init_version_map: Dict[str, dict] = {}
+        self.init_version_map.update(self.targets.get_extension_versions())
+        self.init_version_map.update(self.targets.get_extension_versions(False))
+
+    def analyze_cluster(self, **override_kwargs: dict) -> "ClusterUpgradeState":
+        with Progress(
+            SpinnerColumn("star"),
+            *Progress.get_default_columns(),
+            "Elapsed:",
+            TimeElapsedColumn(),
+            transient=True,
+            disable=bool(self.no_progress),
+        ) as progress:
+            _ = progress.add_task("Analyzing cluster...", total=None)
```

Comment:
can you just do `progress.add_task("Analyzing cluster...", total=None)` - the wildcard is extra

Reply:
I can't remember if it was, or still is, a thing - but I think there was a linter rule regarding calling functions that return a value without assigning that value to something?